### PR TITLE
{Bp-13675} gpio:add feature for gpio wakeup 

### PR DESCRIPTION
**Pull request by jerpelea**
Summary
1.add pin type to cfg gpio as wake up source
2.add IOEXPANDER_OPTION_WAKEUPCFG for set wake up option

Impact
Add more feature in ioexpand.

Testing
CI

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



This pull request appears **incomplete** based on the NuttX requirements. Here's a breakdown:

**What's Missing/Needs Improvement:**

* **Summary:**
    *  **Insufficient Detail:** The summary provides a high-level overview but lacks specifics. 
        * **Why is this change necessary?**  Is it fixing a bug, enhancing functionality, or adding a new feature? Provide context.
        * **How does it work?**  Explain the technical implementation. For example, how is pin type used for the wake-up source? How does the new `IOEXPANDER_OPTION_WAKEUPCFG` option function?
        * **Related Issues:**  Link any relevant NuttX or NuttX Apps issues this PR addresses.

* **Impact:**
    * **User Impact:** While you mention adding a feature, you need to clarify if users need to make any changes to their code or configuration to utilize this new feature.
    * **Build Impact:**  Does this PR introduce any new configuration options, dependencies, or changes to the build process?
    * **Hardware Impact:**  Specify which architectures, boards, or drivers are affected by the IO expander changes. 
    * **Documentation Impact:** State if documentation updates are necessary and whether you have included them in the PR. 
    * **Other Impacts:** The other impact categories are marked "NO" without explanation. If there's genuinely no impact, briefly state that. Otherwise, provide details.

* **Testing:**
    * **Insufficient Information:** "CI" is not enough detail.
    * **Build Host(s):** List the operating system, CPU architecture, and compiler (including versions) used for local testing.
    * **Target(s):**  Specify the architectures, boards, and configurations you tested on. If you used a simulator, mention that.
    * **Testing Logs:** Provide relevant snippets of testing logs that demonstrate the issue before the change and the successful outcome after the change.

**Recommendation:**

Expand each section of your PR description with the missing information and address the points mentioned above. This will make your PR easier for reviewers to understand and approve.

**Review by lupyuen — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by jerpelea** [comment]
@lupyuen 

Warning: nettest_server.c:151:14: warning: implicit declaration of function 'accept4'; did you mean 'accept'? [-Wimplicit-function-declaration]

**Comment by lupyuen** [comment]
> Warning: nettest_server.c:151:14: warning: implicit declaration of function 'accept4'; did you mean 'accept'? [-Wimplicit-function-declaration]

That warning comes from here, it's not an error: https://github.com/apache/nuttx-apps/pull/2595#issuecomment-2381592939

The error is here: https://github.com/apache/nuttx/actions/runs/11143766249/job/30969638637?pr=13792
```text
Error: /github/workspace/sources/apps/system/uorb/sensor/angle.c:39:32: error: invalid application of 'sizeof' to incomplete type 'struct sensor_angle'
   39 | ORB_DEFINE(sensor_hinge_angle, struct sensor_angle, sensor_angle_format);
      |                                ^~~~~~
/github/workspace/sources/apps/system/uorb/uORB/uORB.h:210:12: note: in definition of macro 'ORB_DEFINE'
  210 |     sizeof(structure), \
      |            ^~~~~~~~~
```

Lemme trace it...

**Comment by lupyuen** [comment]
sensor_angle struct comes from here: https://github.com/apache/nuttx/pull/13639

**Comment by lupyuen** [comment]
> can you check that CI runs on the releases/12.7 apps branch instead of master?

Sorry I'm not sure how to check. @xiaoxiang781216 Would you know?

File: Documentation/components/drivers/special/ioexpander.rst [hunk 1]
```diff
@@ -1,3 +1,35 @@
...
+
+-  ``struct ioexpander_ops_s``. Each ioexpand device driver must implement
+   an instance of ``struct ioexpander_ops_s``. That structure defines a
+   call table with the methods, and we also provide macros to help access methods.
+
+-  we also provide method ``gpio_lower_half`` to make ioexpander compatible with normal gpio.
+
+-  **Binding ioexpander Drivers**. ioexpander drivers are not normally directly
+   accessed by user code, we should always get lower level drivers, for example I2C,
+   and map extended gpio feature same asa normal gpio. See for example,
```
Comment (hartmannathan) [inline Documentation/components/drivers/special/ioexpander.rst:17]:
```suggestion
   and map extended gpio feature same as a normal gpio. See for example,
```

Comment (xiaoxiang781216) [reply Documentation/components/drivers/special/ioexpander.rst:17]: this patch come from master branch, let's fix it there and cherry-pick to release branch.

**Pull request merged**